theano/compile/debugmode.py

```diff
@@ -2293,7 +2298,7 @@ def __init__(self, inputs, outputs, optimizer, mode,
                 for output, spec in izip(fgraph.outputs,
                                          outputs + additional_outputs)
                 if not spec.borrow]
-        if no_borrow:
+        if no_borrow and not always_borrow:
```

Comment:
Yes, maybe that could work.  And we don't have shared variables in the inner function so that won't be a problem.